lib/on_form/saving.rb

```diff
@@ -14,26 +14,21 @@ def transaction(&block)
       end
     end
 
-    def invalid?
-      !valid?
-    end
 
     def save!
       reset_errors
       transaction do
         reset_errors
-        unless run_validations!(backing_model_validations: false)
-          backing_model_instances.each(&:valid?)
-          collect_errors
-          raise ActiveModel::ValidationError, self
-        end
+
+        self.backing_model_validations = false
+        has_form_errors = !run_validations!
+        self.backing_model_validations = true
+        run_backing_model_validation!
+        raise ActiveModel::ValidationError, self if has_form_errors
+        raise ActiveRecord::RecordInvalid, self unless errors.empty?
```

Comment:
We got into an interesting discussion here about `ValidationError` vs `RecordInvalid`, and whether there's a meaningful distinction between custom validations failing and model validations failing.

We also talked about when ActiveRecord does particular callbacks in the saving process: it explicitly does validation (with hooks) THEN save (with hooks), so it will happily let you make your model invalid during a before_save callback. We wondered whether it would make sense to mirror this here, so run validations for the model + form, then run save callbacks and validate without saving.

Reply:
IMHO yes there is a meaningful distinction, simply because you wouldn't expect a non-record class to raise a RecordInvalid (what would it set #record to?).

I think if we have to choose, we should usually be aiming to be compatible with ActiveRecord's callback order, unless there is something fatally wrong with it or it is impossible to implement, as it's what people will expect as they refactor fat models into forms.  But I don't think I fully implemented it that way because I think it was difficult to achieve without creating other problems.

BTW, did you notice that the callback order is actually specified in the README?  That would need to be changed if we proceed with this behavior change (and a major version bump, but I expect we'll be doing that anyway).  The readme currently says:

> Note that model save calls are nested inside the form save calls, which means that although form validation takes place before form save starts, model validation takes place after form saving begins.

You mention the ActiveRecord behavior in the case you have both validation callbacks and save callbacks.  What happens in the case where an associated model has both callbacks itself?  Do both the original and associated model before_validations run before the before_save on the original record?